README.md
# primr
 Patient rule-induction method
 
## Description

The `primr` package provides function for performing the patient rule-induction method (PRIM) proposed by Friedman and Fisher (1999). PRIM is designed for bump hunting, i.e. to find a subdomain of `x` inputs in which an objective function of a response `y` is high.

## Installation

1. In R, install the package directly from github using the command (the package `devtools` is required):
```r
> library(devtools)
> install_github("PierreMasselot/primr", build_vignettes = TRUE)
```
2. The package can then be loaded as usual: `library(primr)`.
3. You can see the vignette for simple examples: `vignette("toy_example")`.
4. You can see the list of functions below and access help from R with `?peeling`.

## Functions

The `primr` package revolves around two main functions :
* `peeling`: Performs the top-down peeling consisting by iteratively peeling a box containing the whole dataset such that the objective function increases.
* `pasting`: Refines the final box's edges by slightly expanding it, increasing the objective function value.

Both function produce a `prim` object that contains the peeling trajectory, i.e. the successive peeled boxes. The stopping box of the peeling algorithm can be chosen through different functions:
* `jump.prim`: Selects the stopping box in a `prim` object through a 'jump' criterion.
* `cv.trajectory`: Produces a cross-validated peeling trajectory.
* `plot_trajectory`: Plots the peeling trajectory.

In addition, `prim` objects can be passed to several functions for analysis:
* `extract.box`: Extracts a particular box from a `prim` object.
* `plot_box`: Plots a bidimensional projection of the data with one or several boxes.
* `predict.prim`: For a new set of data, predicts whether each observation falls in the chosen box.

## References

Friedman, J.H., Fisher, N.I., 1999. Bump hunting in high-dimensional data. *Statistics and Computing* 9, 123-143. https://doi.org/10.1023/A:1008894516817

Masselot P., Chebana F., Campagna C., Lavigne E., Ouarda T.B.M.J., Gosselin P. Machine learning approaches to identify thresholds in a heat-health warning system context. *Submitted*.

NEWS.md
## 0.2.2

### Bug fixes
- Bug in peeling with categorical variables.

tests/test_plots.R
library(fields)

x <- matrix(runif(400), ncol = 2)
y <- 10 * x[,1] + 5 * x[,2] + 10 * (x[,1] > .8 & x[,2] > .5) + rnorm(100)

plot(x, col = tim.colors(10)[cut(y, 10)], pch = 16)

object <- peeling.sequence(x = x, y = y)

plot_box(object)
plot_box(object, ypalette = heat.colors(5), box.args = list(lwd = 3))
plot_box(object, support = c(.5, .4, .3), 
  box.args = list(border = terrain.colors(3)))
plot_box(object, select = 1)
plot_box(object, select = 2)
plot_box(object, select = 2, npeel = 1:3*10, 
  box.args = list(border = heat.colors(3)))

plot_trajectory(object)
plot_trajectory(object, type = "b", pch = 16, col = "red")
plot_trajectory(object, xtype = "nobs", col = "blue")
plot_trajectory(object, xtype = "nobs", ytype = "diff")
plot_trajectory(object, xtype = "nobs", ytype = "rel.diff", pch = 15, cex = 0.5)   

tests/testthat/test-pasting.R
test_that("pasting numerical variables works", {
  x <- data.frame(V1 = runif(100))
  y <- 5*x[[1]] + rnorm(100)
  small.box <- list(c(.1, .2))
  nbox <- sum(in.box(x, small.box))
  past_res <- paste.one(y, x, alpha = .2, small.box = small.box,
    obj.fun = construct_objfun(mean))
  # Check that 2 obs have been added
  expect_equal(length(past_res$y), ceiling(nbox * 1.2)) 
  
  # Check that it had been added on the right
  past_res <- paste.one(y, x, alpha = .2, small.box = small.box, 
    peeling.side = 1, obj.fun = construct_objfun(mean))
  expect_equal(past_res$limits[[1]][1], min(x[x >= .1]))
  expect_gt(past_res$limits[[1]][2], .2)
  
  # Check that it had been added on the left
  past_res <- paste.one(y, x, alpha = .2, small.box = small.box, 
    peeling.side = -1, obj.fun = construct_objfun(mean))
  expect_equal(past_res$limits[[1]][2], max(x[x <= .2]))
  expect_lt(past_res$limits[[1]][1], .1)
  
  # Check that it works with several variables
  x <- data.frame(V1 = runif(1000), V2 = runif(1000))
  y <- 5*x[[1]] + rnorm(1000)
  small.box <- list(c(.1, .2), c(.1, .2))
  nbox <- sum(in.box(x, small.box))
  past_res <- paste.one(y, x, alpha = .2, small.box = small.box,
    obj.fun = construct_objfun(mean))
  # Check that 2 obs have been added
  expect_equal(length(past_res$y), ceiling(nbox * 1.2)) 
})

test_that("pasting categorical variables works", {
  set.seed(1)
  x <- data.frame(V1 = sample(letters[1:5], 100, replace = T))
  y <- rnorm(100)
  lettermeans <- aggregate(y, by = x, mean)
  worstletter <- as.character(lettermeans[which.min(lettermeans[,2]),1])
  bestletter <- as.character(lettermeans[which.max(lettermeans[,2]),1])
  past_res <- paste.one(y, x, small.box = list(worstletter), numeric.vars = F,
    obj.fun = construct_objfun(mean))
  # Check that limits now contains the new best letter
  expect_setequal(past_res$limits[[1]], c(worstletter, bestletter))
  expect_gt(length(past_res$y), sum(x == worstletter)) 
})

test_that("sequence pasting works", {
  x <- runif(100)
  y <- -10 * abs(x - .5)
  peeled <- peeling(y, x, peeling.side = 1)
  
  past_res <- pasting(peeled, npeel = peeled$npeel, 
    peeling.side = 0)
  expect_equal(length(past_res$limits), 
    past_res$npaste + peeled$npeel + 1)
  expect_equal(length(past_res$yfun), 
    past_res$npaste + peeled$npeel + 1)
  expect_equal(length(past_res$support), 
    past_res$npaste + peeled$npeel + 1)
  expect_gte(past_res$support[past_res$npaste + peeled$npeel + 1], 
    peeled$support[peeled$npeel+ 1])

    # Check robustness to misspecification of the chosen box
  expect_warning(
    past_res <- pasting(peeled, npeel = peeled$npeel, support = 0.5))
  expect_length(past_res$support, past_res$npeel + past_res$npaste + 1)
})

tests/testthat/test-peeling.R
test_that("peeling numeric variables works", {
  x <- runif(100)
  y <- 5*x + rnorm(100)
  
  # General peeling
  peel_res <- peel(y, as.matrix(x), limits = list(c(0,1)),
    obj.fun = construct_objfun(mean))
  expect_equal(peel_res$support, 0.95)
  
  # Left peeling
  peel_res <- peel(y, as.data.frame(x), limits = list(c(0,1)), peeling.side = -1,
    obj.fun = construct_objfun(mean))
  expect_equal(peel_res$limits[[1]][1], sort(x)[6])
  expect_equal(peel_res$limits[[1]][2], max(x))
  
  # Right peeling
  peel_res <- peel(y, as.data.frame(x), limits = list(c(0,1)), peeling.side = 1,
    obj.fun = construct_objfun(mean))
  expect_equal(peel_res$limits[[1]][2], sort(x)[95])
  expect_equal(peel_res$limits[[1]][1], min(x))
  
  # Peeling two variables
  x <- as.data.frame(matrix(runif(200), ncol = 2))
  xlims <- lapply(x, range)
  y <- 1000*x[,1] + rnorm(100)
  
  peel_res <- peel(y, x, limits = xlims,
    obj.fun = construct_objfun(mean))
  xlims[[1]][1] <- sort(x[,1])[6]
  expect_equal(peel_res$limits, xlims)
})

test_that("peeling sequence works", {
  x <- matrix(runif(200), ncol = 2)
  y <- 100 * x[,1] + rnorm(100)
  
  peel_res <- peeling(y, x)
  expect_gte(peel_res$yfun[peel_res$npeel], mean(y))
  expect_true(all(diff(peel_res$support) < 0))
  x1lims <- sapply(peel_res$limits, "[[", 1)
  expect_true(all(diff(x1lims[1,]) >= 0))
  expect_true(all(diff(x1lims[2,]) <= 0))
  x2lims <- sapply(peel_res$limits, "[[", 2)
  expect_true(all(diff(x2lims[1,]) >= 0))
  expect_true(all(diff(x2lims[2,]) <= 0))
})

R/plots.R
###############################################################################
#
#                      Plot functions                    
#
###############################################################################

#' Plot a box
#'
#' Plots a bidimensional projection of the data and chosen boxes.
#'
#' @param object A 'prim' object.
#' @param select A vector of length 1 or 2 indicating the input variables(s) 
#'    to plot. Default to the two first ones. 
#'    If a single variable is selected, plot it against \code{y}.
#' @param npeel Integer vector indicating the number of peeling iteration
#'    of boxes to plot.
#' @param support Numeric vector with values between 0 and 1 indicating the 
#'    support of boxes to plot.
#' @param yfun Numeric vector indicating the value of the objective function
#'    of the boxes to plot.
#' @param npaste Integer vector indicating the number of pasting iteration
#'    of boxes to plot.
#' @param ypalette A palette of colors for representing the \code{y} value
#'    associated to each point.
#' @param box.args A list of arguments to be passed to 
#'    \code{\link[graphics]{rect}} for drawing boxes. All arguments can be 
#'    given as vectors to specify different values for each drawn box. 
#' @param ... Additional graphical parameters for 
#'    \code{\link[graphics]{plot}}.
#'
#' @details Several boxes can be displayed at once, selected by one of the
#'    arguments \code{npeel}, \code{support}, \code{yfun} or \code{npaste}
#'    (when relevant). Note that the arguments in \code{box.args} allow
#'    giving different parameters to each displayed box.
#'
#' @seealso \code{\link{peeling}} for peeling trajectories. 
#'
#' @examples
#'    # A simple bump
#'    set.seed(12345)
#'    x <- matrix(runif(2000), ncol = 2, dimnames = list(NULL, c("x1", "x2")))
#'    y <- 2 * x[,1] + 5 * x[,2] + 10 * (x[,1] >= .8 & x[,2] >= .5) + 
#'      rnorm(1000)
#'    # Peeling with alpha = 0.05 and beta.stop = 0.05
#'    peel_res <- peeling(y, x, beta.stop = 0.05)
#'    # Automatically choose the best box
#'    chosen <- jump.prim(peel_res)
#'    # Plot the resulting box
#'    plot_box(peel_res, pch = 16, ypalette = hcl.colors(10), 
#'      support = chosen$final.box$support, box.args = list(lwd = 2))
#'
#' @export
plot_box <- function(object, select = 1:2, npeel = NULL, support = NULL, 
  yfun = NULL, npaste = NULL, ypalette = NULL, box.args = list(), ...)
{
  dots <- list(...)
  if (!inherits(object, "prim")){
    warning("'object' should be of class 'prim'")
  } 
  n <- nrow(object$x)
  p <- ncol(object$x)
  # Select the x to plot
  select <- select[1:pmin(length(select), 2, p)]
  xplot <- object$x[,select, drop = FALSE]
  labs <- colnames(xplot)
  # Extract boxes
  boxes <- extract.box(object, npeel = npeel, support = support, 
    yfun = yfun, npaste = npaste)
  nboxes <- length(boxes$yfun)
  boxlimits <- lapply(boxes$limits, "[", select)
  # If a single dimension is to be plotted
  if (length(select) == 1){
    xplot <- cbind(xplot, object$y, deparse.level = 0)
    labs[2] <- "y"
    ylims <- range(object$y) + diff(range(object$y)) * c(-0.02, 0.02)
    boxlimits <- lapply(boxlimits, append, list(ylims))
  }
  # Preparing default parameters
  if (!is.null(ypalette)){
    cols <- ypalette[cut(object$y, length(ypalette))]
  } else {
    cols <- "black"
  }
  dots.def <- list(xlab = labs[1], ylab = labs[2], col = cols)
  plot.args <- c(list(x = xplot), dots, 
    dots.def[!names(dots.def) %in% names(dots)])
  do.call(graphics::plot, plot.args)
  # Add boxes
  box.args <- lapply(box.args, rep_len, nboxes)
  box.args$xleft <- sapply(boxlimits, function(x) x[[1]][1])
  box.args$xright <- sapply(boxlimits, function(x) x[[1]][2])
  box.args$ybottom <- sapply(boxlimits, function(x) x[[2]][1])
  box.args$ytop <- sapply(boxlimits, function(x) x[[2]][2])
  box.args$f <- graphics::rect
  invisible(do.call(Map, box.args)) 
}



#' Plot a peeling trajectory
#'
#' Displays the peeling trajectory of a \code{prim} object with chosen
#'    cut-points.
#'
#' @param object A \code{prim} object.
#' @param xtype A character indicating the how to display the x axis of
#'    the plot. 'support' (the default) for the support of the boxes or
#'    'nobs' for the number of observations inside the box.
#' @param ytype A character indicating which trajectory to plot. 'yfun'
#'    (the default) for the objective function value, 'diff' for the
#'    difference between successive boxes and 'rel.diff' for the relative
#'    difference. See \code{\link{jump.prim}} for details.
#' @param npeel Integer vector. If not \code{NULL}, draw a vertical line 
#'    at the corresponding peeling iterations of the trajectory.
#' @param support Numeric vector. If not \code{NULL}, draw a vertical line 
#'    at the corresponding supports of the trajectory.
#' @param yfun Numeric vector. If not \code{NULL}, draw a vertical line 
#'    at the boxes with the closest \code{yfun} value.
#' @param npaste Integer vector. If not \code{NULL}, draw a vertical line 
#'    at the corresponding pasting iterations of the trajectory.
#' @param abline.pars List of parameters to be passed to
#'    \code{\link[graphics]{abline}} for the vertical lines.
#' @param se Logical indicating if standard error bars should be drawn.
#'    Works only for \code{cv.prim} objects and \code{ytype = "yfun"}.
#' @param se.pars Parameters to be passed to \code{arrows} for drawing
#'    strandard error bars.
#' @param ... Additional graphical parameters for 
#'    \code{\link[graphics]{plot}}.
#'
#' @seealso \code{\link{peeling}} for peeling trajectories. 
#'
#' @references
#'    Friedman, J.H., Fisher, N.I., 1999. Bump hunting in high-dimensional data.
#'      Statistics and Computing 9, 123-143. 
#'      https://doi.org/10.1023/A:1008894516817
#'
#' @examples
#'    # A simple bump
#'    set.seed(12345)
#'    x <- matrix(runif(2000), ncol = 2, dimnames = list(NULL, c("x1", "x2")))
#'    y <- 2 * x[,1] + 5 * x[,2] + 10 * (x[,1] >= .8 & x[,2] >= .5) + 
#'      rnorm(1000)
#'    # Peeling with alpha = 0.05 and beta.stop = 0.05
#'    peel_res <- peeling(y, x, beta.stop = 0.05)
#'
#'    # Display the peeling trajectory
#'    plot_trajectory(peel_res, type = "b", pch = 16, col = "cornflowerblue", 
#'      support = 0.11, abline.pars = list(lwd = 2, col = "indianred"), 
#'      xlab = "")
#'
#' @export
plot_trajectory <- function(object, xtype = c("support", "nobs"), 
  ytype = c("yfun", "diff", "rel.diff"), npeel = NULL, support = NULL, 
  yfun = NULL, npaste = NULL, abline.pars = list(), 
  se = TRUE, se.pars = list(), ...)
{
  dots <- list(...)
  if (!inherits(object, c("prim", "cv.prim"))){
    warning("'object' should be of class 'prim' or 'cv.prim'")
  }
  n <- nrow(object$x)
  nbox <- object$npeel + 1
  dots.def <- list()
  if (length(c(npeel, support, yfun, npaste)) > 0){
    boxes <- extract.box(object, npeel = npeel, support = support, 
      yfun = yfun, npaste = npaste)
  } else {
    boxes <- NULL
  }
  ab.def <- list(v = boxes$support, lty = 2)
  xtype <- match.arg(xtype) 
  switch(xtype, 
    support = {
      x <- object$support
      dots.def$xlab <- "Support"
      ab.def$v <- boxes$support
    },
    nobs = {
      x <- round(object$support * n)
      dots.def$xlab <- "Number of observations"
      ab.def$v <- round(boxes$support * n)
    }
  )
  ytype <- match.arg(ytype)
  switch(ytype, 
    yfun = {
      y <- object$yfun
      dots.def$ylab <- "Objective function"
    },
    diff = {
      y <- jump.prim(object, rel.support = F)$trajectory.difference
      x <- x[-1]
      dots.def$ylab <- "Difference"
    },
    rel.diff = {
      y <- jump.prim(object, rel.support = T)$trajectory.difference
      x <- x[-1]
      dots.def$ylab <- "Relative difference"
    }
  )
  plot.args <- c(list(x = x, y = y), dots, 
    dots.def[!names(dots.def) %in% names(dots)])
  do.call(graphics::plot, plot.args)
  if (se && inherits(object, "cv.prim")){
    se.def <- list(length = 0, col = "grey")
    se.pars <- c(list(x0 = x, y0 = y, 
      y1 = c(y + object$se.yfun, y - object$se.yfun)), 
      se.pars, se.def[!names(se.def) %in% names(se.pars)])
    do.call(graphics::arrows, se.pars)
  }
  if (length(c(npeel, support, yfun)) > 0){
    abline.pars <- c(abline.pars, 
      ab.def[!names(ab.def) %in% names(abline.pars)])
    do.call(graphics::abline, abline.pars)
  }  
}



R/util_functions.R
#' Construct an objective function from a user given one.
#'
#' @param fun The function given to \code{obj.fun} in the 
#'    main function.
#'
#' @details The purpose is to create an objective function
#'    that takes two arguments: \code{y} for the response
#'    vector and \code{inbox} a logical vector indicating
#'    if each observation lies in the box for which the
#'    objective function is computed.
construct_objfun <- function(fun){
  fun <- match.fun(fun)
  arguments <- names(formals(fun))
  if (!all(c("y", "x", "inbox") %in% arguments)){
    if (!all(c("y", "x") %in% arguments)){
      objfun <- function(y, x = NULL, inbox = T){
        y <- y[inbox]
        do.call(fun, list(y))
      }
    } else {
      objfun <- function(y, x = NULL, inbox = T){
        y <- y[inbox]
        x <- x[inbox,]
        do.call(fun, list(y, x))
      }
    }
  } else {
    objfun <- fun
  }
  return(objfun)
}

R/peeling.R
###############################################################################
#
#                   Peeling functions                             
#
###############################################################################

#' Top-down peeling
#'
#' Iteratively peels a dataset for bump hunting.
#'
#' @param y Numeric vector of response values.
#' @param x Numeric or categorical data.frame of input values.
#' @param alpha The peeling fraction of the algorithm. A value between 0 and 1
#'    giving the proportion of peeled observations at each step.
#' @param beta.stop The stopping support of the algorithm. A value 
#'    between 0 and 1 giving the proportion of remaining data below
#'    which the algorithm stops.
#' @param obj.fun The function of \code{y} to be maximized. Can be a user 
#'    defined function (see details).
#' @param peeling.side A numeric vector for side constraints on the peeling of 
#'    each input variable. -1 indicates peeling only the 'left' of the box
#'    (i.e. increasing the lower limit only), 1 indicate peeling only the
#'    'right' and 0 for no constraint.
#'
#' @details The function \code{peeling} carries out the top-down peeling 
#'    which is the first step of the PRIM algorithm. At each iteration 
#'    it peels a proportion \code{alpha} of data from one side of the domain
#'    in order to increase the value of the function \code{obj.fun} applied
#'    to the response \code{y}. The algorithm iterates the peeling until 
#'    the support of the box (i.e. the proportion of remaining observations)
#'    is below the value \code{beta.stop}.
#'
#'    Many function can be used in \code{obj.fun} including user defined 
#'    functions. User defined function should take two arguments: \code{y}
#'    and \code{x} representing corresponding variables 
#'    and \code{inbox} which is a boolean
#'    vector indicating the observations inside the current box.
#'    Note that a classical function can also be passed to \code{obj.fun}
#'    such as \code{mean}, \code{var} or \code{median}. In this case
#'    the function is created internally to fit the above structure. 
#'    For more functions more complicated than the basic ones,
#'    it is recommended that the user set its own function as stated
#'    above.  
#'    
#'    The function also allows directed peeling, i.e. to contraint the peeling
#'    occuring on a single side of some input variables. Thus when 
#'    \code{peeling.side = -1}, only the lower part of the variable is peeled
#'    (the "left" of the domain) and when \code{peeling.side = 1}, only the
#'    upper part of the variable is peeled. Note that a vector can be passed,
#'    thus applying different constraints to the input variables.
#'
#' @return A \code{prim} object which is a list with the following elements:
#'    \item{npeel}{The number of peeling iteration performed.}
#'    \item{support}{A vector of length \code{npeel + 1} containing the support 
#'      of each successivepeeled box.}
#'    \item{yfun}{A vector of length \code{npeel + 1} containing the objective 
#'      function value of each successive peeled box.}
#'    \item{limits}{A list of length \code{npeel + 1} containing the limits 
#'      of each successive box. Each limit is a list with one element per input 
#'      variable.}
#'    \item{x,y}{The input and response data used in the algorithm.}
#'    \item{numeric.vars}{A logical vector indicating, for each input variable,
#'      if it was considered as a numeric variable.}
#'    \item{alpha, peeling.side, obj.fun}{The value of the arguments used for
#'      peeling. Useful for prim methods.}
#'    \item{npaste}{Number of pasting iteration performed. Should be 0 here,
#'      but useful for \code{\link{pasting}}.}
#'
#'    Note that the first box in a \code{prim} object is the starting box
#'      containing the whole dataset. This is why the \code{limits},
#'      \code{yfun} and \code{support} elements have length \code{npeel + 1}.
#'
#' @seealso \code{\link{extract.box}} to extract information about a 
#'    particular box in a \code{prim} object. 
#'    \code{\link{plot_trajectory}} and \code{\link{plot_box}} to explore
#'    the peeling trajectory. \code{\link{jump.prim}} to automatically
#'    choose the best box. \code{\link{predict.prim}} to predict if new data
#'    falls into particular boxes. \code{\link{pasting}} to carry out the
#'    pasting refining the edges of the chosen box.
#'
#' @references
#'    Friedman, J.H., Fisher, N.I., 1999. Bump hunting in high-dimensional data.
#'      Statistics and Computing 9, 123-143. 
#'      https://doi.org/10.1023/A:1008894516817
#'
#' @examples
#'    # A simple bump
#'    set.seed(12345)
#'    x <- matrix(runif(2000), ncol = 2, dimnames = list(NULL, c("x1", "x2")))
#'    y <- 2 * x[,1] + 5 * x[,2] + 10 * (x[,1] >= .8 & x[,2] >= .5) + 
#'      rnorm(1000)
#'    # Peeling with alpha = 0.05 and beta.stop = 0.05
#'    peel_res <- peeling(y, x, beta.stop = 0.05)
#'    # Automatically choose the best box
#'    chosen <- jump.prim(peel_res)
#'    # Plot the resulting box
#'    plot_box(peel_res, pch = 16, ypalette = hcl.colors(10), 
#'      support = chosen$final.box$support, box.args = list(lwd = 2))
#'
#'    # Examples of directed peeling
#'    set.seed(12345)
#'    x <- matrix(runif(2000), ncol = 2, dimnames = list(NULL, c("x1", "x2")))
#'    y <- 10 * (x[,1] <= .2 & x[,2] <= .2) + 10 * (x[,1] >= .8 & x[,2] >= .8) +
#'      rnorm(1000)
#'    # Left peeling
#'    peel_left <- peeling(y, x, peeling.side = -1)
#'    chosen <- jump.prim(peel_left)
#'    plot_box(peel_left, pch = 16, ypalette = hcl.colors(10), 
#'      support = chosen$final.box$support, box.args = list(lwd = 2),
#'      main = "Left peeling")
#'    # Right peeling
#'    peel_right <- peeling(y, x, peeling.side = 1)
#'    chosen <- jump.prim(peel_right)
#'    plot_box(peel_right, pch = 16, ypalette = hcl.colors(10), 
#'      support = chosen$final.box$support, box.args = list(lwd = 2),
#'      main = "Right peeling")
#'
#'    # User-defined objective function to minimize the mean
#'    set.seed(3333)
#'    x <- matrix(runif(2000), ncol = 2, dimnames = list(NULL, c("x1", "x2")))
#'    y <- - 10 * (x[,1] <= .2 & x[,2] <= .2) + 10 * (x[,1] >= .8 & x[,2] >= .8) +
#'      rnorm(1000)
#'    peel_res <- peeling(y, x, obj.fun = function(x) -mean(x))
#'    chosen <- jump.prim(peel_res)
#'    plot_box(peel_res, pch = 16, ypalette = hcl.colors(10), 
#'      support = chosen$final.box$support, box.args = list(lwd = 2))
#'
#'    # User-defined function maximizing the slope of a linear regression
#'    set.seed(5555)
#'    x <- runif(500)
#'    ym <- 0.5 * x + 5 * (x - 0.7) * (x >= 0.7)
#'    y <- ym + rnorm(500, sd = 0.1)    
#'    peel_res <- peeling(y, x, beta.stop = 0.1, 
#'      obj.fun = function(y, x, inbox){
#'        dat <- data.frame(y, x)
#'        coef(lm(y ~ x, data = dat[inbox,]))[2]
#'    })   
#'    par(mfrow = c(1,2))
#'    plot_trajectory(peel_res, type = "b", pch = 16, col = "cornflowerblue", 
#'      support = 0.3, abline.pars = list(lwd = 2, col = "indianred"))
#'    plot_box(peel_res, pch = 16, ypalette = hcl.colors(10), 
#'      support = 0.3, box.args = list(lwd = 2))
#'    lines(sort(x), ym[order(x)], col = "red", lwd = 2)
#'    
#' @export
peeling <- function(y, x, alpha = 0.05, beta.stop = 0.01, 
  obj.fun = mean, peeling.side = 0)  
{
    x <- as.data.frame(x)
    p <- ncol(x)
    n <- nrow(x)
    numeric.vars <- sapply(x, is.numeric)
    nnumeric <- sum(numeric.vars)
    x[!numeric.vars] <- lapply(x[!numeric.vars], as.factor)
    peeling.side <- rep_len(peeling.side, nnumeric)
    nstep <- ceiling(log(beta.stop)/log(1-alpha))
    support <- yfun <- numeric(nstep + 1)
    limits <- vector("list", nstep+1)
    support[1] <- 1    
    obj.fun <- construct_objfun(obj.fun)
    yfun[1] <- do.call(obj.fun, list(y = y, x = x, inbox = rep(T, n)))
    limits[[1]] <- vector("list", p)
    limits[[1]][numeric.vars] <- lapply(x[,numeric.vars, drop = F], range)
    limits[[1]][!numeric.vars] <- lapply(x[,!numeric.vars, drop = F], unique)
    count <- 1
    while (support[count] > beta.stop){
      new.peel <- peel(y = y, x = x, alpha = alpha, obj.fun = obj.fun, 
        peeling.side = peeling.side, limits = limits[[count]], 
        numeric.vars = numeric.vars)
      count <- count + 1
      support[count] <- new.peel$support
      yfun[count] <- new.peel$yfun
      limits[[count]] <- new.peel$limits
      if (count == (nstep + 1) || 
        length(unique(y[in.box(x, limits[[count]])])) == 1){ 
          break
      }
    }
    if (count < nstep){
      irem <- count:nstep + 1
      support <- support[-irem]
      yfun <- yfun[-irem]
      limits <- limits[-irem]
    }  
    out <- list(npeel = count - 1, support = support, yfun = yfun, 
      limits = limits, x = x, y = y, numeric.vars = numeric.vars, alpha = alpha, 
      peeling.side = peeling.side, obj.fun = obj.fun,
      npaste = 0)
    class(out) <- "prim"
    return(out)
}

peel <- function(y, x, alpha = 0.05, obj.fun = mean, limits, 
  numeric.vars = rep(TRUE, ncol(x)), peeling.side = rep(0,sum(numeric.vars)))
{
    p <- ncol(x)
    n <- nrow(x)
    yfun <- -Inf
    nnumeric <- sum(numeric.vars)
    numinds <- which(numeric.vars)
    inbox <- in.box(x, limits)
    yin <- y[inbox]
    xin <- x[inbox,, drop = FALSE]
    for (j in 1:p){
      if (numeric.vars[j]){
        boxes <- list(c(alpha, 1), c(0, 1 - alpha))
        boxes <- boxes[peeling.side[numinds == j] != c(1, -1)]
      } else {
        if (length(limits[[j]]) > 1){
          boxes <- limits[[j]]
        } else {
          next
        }
      }            
      for (k in 1:length(boxes)){
        if (numeric.vars[j]){
          newlims <- stats::quantile(xin[,j], boxes[[k]])
          movinglim <- which(boxes[[k]] %in% c(alpha,1 - alpha))
          if (newlims[movinglim] == limits[[j]][movinglim]){ 
          # To manage the case in which there are many ties 
            newlims[movinglim] <- sort(unique(xin[,j]), 
              decreasing = as.logical(movinglim - 1))[2]
          }
          inboxk <- x[,j] >= newlims[1] & x[,j] <= newlims[2]
        } else {
          inboxk <- x[,j] != boxes[k]
          newlims <- limits[[j]][-k]
        }
        jyfun <- do.call(obj.fun, list(y = y, x = x, inbox = inbox & inboxk))
        if (jyfun > yfun){
           finbox <- inbox & inboxk
           yfun <- jyfun
        }
      }      
    }
    # Final readjustment of limits
    final.x <- x[finbox,,drop = FALSE] 
    new.limits <- Map(function(xj, numj){
      if (numj){
        out <- range(xj)
      } else {
        out <- unique(xj)
      }
    }, final.x, numeric.vars)
    return(list(limits = new.limits, yfun = yfun, support = mean(finbox)))
}

